rgw multisite: automated mdlog trimming

cbodley opened the pull request:
implements the coordinated trimming of mdlog entries once they're no longer needed. the algorithm is split between the metadata master zone, which is authoritative for determining which entries are safe to trim, and the non-master peer zones, which only trim up to the master's trim position

the metadata master zone decides which entries are safe to trim by querying the metadata sync status of all peer zones. an mdlog entry is only safe to trim if all peers indicate a sync status marker greater than or equal to that entry. similarly, entire mdlog periods can be safely removed once all peer zones report a `realm_epoch` higher than that mdlog period. this logic guarantees that no mdlog entries will be trimmed while other zones may try to consume them

the non-master peer zones trim mdlog entries based on the trim info they read from the master zone, which consists of an array of timestamps for each shard of the mdlog. timestamps are used here instead of markers because peers can only trip -up to- the master's oldest entry, so the master reports its oldest timestamp minus a small delta (1ms). old mdlog periods can be removed based on the `oldest_realm_epoch` reported in the master's mdlog info. this logic guarantees that peers will not trim mdlog entries that are still present on the master, in case that zone is later promoted to master and expected to serve those entries to other zones

as each old mdlog period is removed, it updates the `RGWMetadataLogHistory` in the `meta.history` object (as reported by `RGWOp_MDLog_Info` in the /admin/log api). this update is made atomic with the use of `RGWObjVersionTracker`, and required adding `objv_tracker` arguments to `RGWSimpleRadosRead/WriteCR` and `put_system_obj_data()` in `RGWRados` and `RGWCache`

the mdlog trim process runs every `rgw_sync_log_trim_interval` (default: 20min) in the same 
`RGWSyncLogTrimThread` used for datalog trimming. to coordinate between multiple gateways in the same zone, a rados lock is held on the `meta.history` object during trim to avoid duplicating the work. this polling and locking logic is encapsulated in class `MetaTrimPollCR`

(this branch depends on https://github.com/ceph/ceph/pull/13067 and https://github.com/ceph/ceph/pull/13070, which are included here as the first 5 commits)

TODO:
- [X] add `radosgw-admin` command to run this sync logic (without requiring lease)
- [X] add `test_multi.py` test to verify output of `radosgw-admin mdlog list` after trim

cbodley commented:
squashed small fixes, added `radosgw-admin mdlog autotrim` command, and added trim checks to `test_multi_period_incremental_sync()` in test_multi.py

cbodley commented:
passed teuthology at http://pulpito.ceph.com/cbodley-2017-02-03_15:59:33-rgw-wip-cbodley-testing---basic-mira/

(2 valgrind issues were in the osd, and the one s3test failure was due to requests taking slightly > 30s)

yehudasa commented on src/rgw/rgw_cr_rados.h:
```diff
@@ -307,17 +310,16 @@ class RGWSimpleRadosWriteCR : public RGWSimpleCoroutine {
-                                                req(NULL) {
+		      const T& _data, RGWObjVersionTracker *objv_tracker = nullptr)
+    : RGWSimpleCoroutine(_store->ctx()), async_rados(_async_rados),
+      store(_store), pool(_pool), oid(_oid), objv_tracker(objv_tracker) {
```
@cbodley lost req initialization

yehudasa commented on src/rgw/rgw_metadata.cc:
```diff
@@ -369,6 +369,128 @@ int write_history(RGWRados *store, const RGWMetadataLogHistory& state,
+  RGWObjVersionTracker *objv;
+  RGWMetadataLogHistory state;
+ public:
+  WriteHistoryCR(RGWRados *store, Cursor cursor, RGWObjVersionTracker *objv)
```
@cbodley why not pass cursor as a const ref?

yehudasa commented on src/rgw/rgw_sync.cc:
```diff
@@ -2226,3 +2226,124 @@ int RGWCloneMetaLogCoroutine::state_store_mdlog_entries_complete()
 }
 
 
+#undef dout_prefix
```
@cbodley maybe it's better to put this code in a separate .cc file?

yehudasa commented on src/rgw/rgw_sync.cc:
```diff
@@ -2226,3 +2226,124 @@ int RGWCloneMetaLogCoroutine::state_store_mdlog_entries_complete()
+        using StatusCR = RGWReadRESTResourceCR<rgw_meta_sync_status>;
+        auto conn = c.second.get();
+        spawn(new StatusCR(cct, conn, env.http, "/admin/log/", params, &*p),
+              false);
```
@cbodley some other place we have a window of operations, so that we don't spawn more than X concurrent crs at once. Maybe here is a good candidate for doing the same?

cbodley commented:
pushed an update:
* rebased, resolving conflicts with rgw_raw_obj and test_multi.py
* removed prerequisite commits that merged in https://github.com/ceph/ceph/pull/13067 and https://github.com/ceph/ceph/pull/13070
* addressed review comments (except for splitting trim into separate source files - i added `// TODO: move into rgw_sync_trim.cc` for now)

consistently passing trim test in `test_multi_period_incremental_sync()`. will get another run through teuthology

cbodley commented:
passed rgw suite: http://pulpito.ceph.com/cbodley-2017-04-27_23:02:56-rgw-wip-cbodley-testing---basic-mira/

cbodley commented:
@yehudasa should be ready to merge. anything else you'd like to see here?

cbodley commented:
ping @yehudasa
